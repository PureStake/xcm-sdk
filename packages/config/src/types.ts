export type AssetId = number | bigint | string;
