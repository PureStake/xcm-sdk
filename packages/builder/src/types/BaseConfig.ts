import { CallType } from '../builder.interfaces';

export interface BaseConfigConstructorParams {
  module: string;
  func: string;
  type: CallType;
}

export class BaseConfig {
  readonly module: string;

  readonly func: string;

  readonly type: CallType;

  constructor({ module, func, type }: BaseConfigConstructorParams) {
    this.module = module;
    this.func = func;
    this.type = type;
  }
}
