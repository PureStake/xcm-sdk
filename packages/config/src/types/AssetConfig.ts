import {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import { AnyChain, Asset } from '@moonbeam-network/xcm-types';

export interface AssetConfigConstructorParams {
  asset: Asset;
  balance: BalanceConfigBuilder;
  contract?: ContractConfigBuilder;
  destination: AnyChain;
  destinationFee: DestinationFeeConfig;
  extrinsic?: ExtrinsicConfigBuilder;
  fee?: FeeAssetConfig;
  min?: AssetMinConfigBuilder;
}

export interface DestinationFeeConfig extends FeeAssetConfig {
  amount: number | FeeConfigBuilder;
}

export interface FeeAssetConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  xcmDeliveryFeeAmount?: number;
}

export class AssetConfig {
  readonly asset: Asset;

  readonly balance: BalanceConfigBuilder;

  readonly contract?: ContractConfigBuilder;

  readonly destination: AnyChain;

  readonly destinationFee: DestinationFeeConfig;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  readonly fee?: FeeAssetConfig;

  readonly min?: AssetMinConfigBuilder;

  constructor({
    asset,
    balance,
    contract,
    destination,
    destinationFee,
    extrinsic,
    fee,
    min,
  }: AssetConfigConstructorParams) {
    this.asset = asset;
    this.balance = balance;
    this.contract = contract;
    this.destination = destination;
    this.destinationFee = destinationFee;
    this.extrinsic = extrinsic;
    this.fee = fee;
    this.min = min;
  }
}
