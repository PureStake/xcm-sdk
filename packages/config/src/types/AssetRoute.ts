import type {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
  MrlConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import type { AnyChain, Asset, SetOptional } from '@moonbeam-network/xcm-types';

export interface AssetRouteConstructorParams {
  asset: Asset;
  source: SourceConfig;
  destination: DestinationConfig;
  contract?: ContractConfigBuilder;
  extrinsic?: ExtrinsicConfigBuilder;
  mrl?: MrlConfig;
}

export interface SourceConfig {
  chain: AnyChain;
  balance: BalanceConfigBuilder;
  fee?: FeeConfig;
  destinationFee?: {
    balance: BalanceConfigBuilder;
  };
  min?: AssetMinConfigBuilder;
}

export interface DestinationConfig extends Omit<SourceConfig, 'fee'> {
  fee: DestinationFeeConfig;
}

export interface FeeConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  // NOTE: Sometimes we need to add some extra amount ("XCM Delivery Fee") to a fee
  // that is returned by "paymentInfo" for extrinsic to not fail.
  extra?: number;
}

export interface MrlConfig {
  isAutomatic: boolean;
  transfer: MrlConfigBuilder;
  moonChainFee: MoonChainFeeConfig;
}

export interface DestinationFeeConfig
  extends SetOptional<FeeConfig, 'balance'> {
  amount: number | FeeConfigBuilder;
}

export interface MoonChainFeeConfig extends FeeConfig {
  amount: number | FeeConfigBuilder;
}

export class AssetRoute {
  readonly asset: Asset;

  readonly source: SourceConfig;

  readonly destination: DestinationConfig;

  readonly contract?: ContractConfigBuilder;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  readonly mrl?: MrlConfig;

  constructor({
    asset,
    source,
    destination,
    contract,
    extrinsic,
    mrl,
  }: AssetRouteConstructorParams) {
    this.asset = asset;
    this.source = source;
    this.destination = destination;
    this.contract = contract;
    this.extrinsic = extrinsic;
    this.mrl = mrl;
  }
}
