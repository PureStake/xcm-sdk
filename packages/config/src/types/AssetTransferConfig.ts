import {
  AssetMinConfigBuilder,
  BalanceConfigBuilder,
  ContractConfigBuilder,
  ExtrinsicConfigBuilder,
  FeeConfigBuilder,
} from '@moonbeam-network/xcm-builder';
import { AnyChain, Asset } from '@moonbeam-network/xcm-types';

export interface AssetTransferConfigConstructorParams {
  asset: Asset;
  balance: BalanceConfigBuilder;
  contract?: ContractConfigBuilder;
  destination: AnyChain;
  destinationFee: DestinationFeeConfig;
  extrinsic?: ExtrinsicConfigBuilder;
  fee?: FeeAssetConfig;
  min?: AssetMinConfigBuilder;
}

export interface DestinationFeeConfig extends FeeAssetConfig {
  amount: number | FeeConfigBuilder;
}

export interface FeeAssetConfig {
  asset: Asset;
  balance: BalanceConfigBuilder;
  // Sometimes we need to add some extra amount ("XCM Delivery Fee") to a fee that is returned by "paymentInfo" for extrinsic to not fail.
  extra?: number;
}

export class AssetTransferConfig {
  readonly asset: Asset;

  readonly balance: BalanceConfigBuilder;

  readonly contract?: ContractConfigBuilder;

  readonly destination: AnyChain;

  readonly destinationFee: DestinationFeeConfig;

  readonly extrinsic?: ExtrinsicConfigBuilder;

  readonly fee?: FeeAssetConfig;

  readonly min?: AssetMinConfigBuilder;

  constructor({
    asset,
    balance,
    contract,
    destination,
    destinationFee,
    extrinsic,
    fee,
    min,
  }: AssetTransferConfigConstructorParams) {
    this.asset = asset;
    this.balance = balance;
    this.contract = contract;
    this.destination = destination;
    this.destinationFee = destinationFee;
    this.extrinsic = extrinsic;
    this.fee = fee;
    this.min = min;
  }
}
