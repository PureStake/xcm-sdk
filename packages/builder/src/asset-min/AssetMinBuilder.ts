/* eslint-disable @typescript-eslint/no-use-before-define */

import { Option } from '@polkadot/types';
import { PalletAssetsAssetDetails } from '@polkadot/types/lookup';
import { SubstrateQueryConfig } from '../types/substrate/SubstrateQueryConfig';
import { AssetMinConfigBuilder } from './AssetMinBuilder.interfaces';

export function AssetMinBuilder() {
  return {
    assetRegistry,
    assets,
  };
}

function assetRegistry() {
  const pallet = 'assetRegistry';
  return {
    assetMetadatas: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: pallet,
          func: 'assetMetadatas',
          args: [asset],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          transform: async (response: Option<any>): Promise<bigint> =>
            response.unwrapOrDefault().minimalBalance.toBigInt(),
        }),
    }),
    currencyMetadatas: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: pallet,
          func: 'currencyMetadatas',
          args: [asset],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          transform: async (response: Option<any>): Promise<bigint> =>
            response.unwrapOrDefault().minimalBalance.toBigInt(),
        }),
    }),
  };
}

function assets() {
  return {
    asset: (): AssetMinConfigBuilder => ({
      build: ({ asset }) =>
        new SubstrateQueryConfig({
          module: 'assets',
          func: 'asset',
          args: [asset],
          transform: async (
            response: Option<PalletAssetsAssetDetails>,
          ): Promise<bigint> =>
            response.unwrapOrDefault().minBalance.toBigInt(),
        }),
    }),
  };
}
