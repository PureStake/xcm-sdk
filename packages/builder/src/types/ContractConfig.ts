/* eslint-disable @typescript-eslint/no-explicit-any */
import { BaseConfig, BaseConfigConstructorParams } from './BaseConfig';

export interface ContractConfigConstructorParams
  extends BaseConfigConstructorParams {
  address?: string;
  args: any[];
}

export class ContractConfig extends BaseConfig {
  readonly address?: string;

  readonly args: any[];

  static is(obj: unknown): obj is ContractConfig {
    return obj instanceof ContractConfig;
  }

  constructor({ args, address, ...other }: ContractConfigConstructorParams) {
    super({ ...other });

    this.args = args;
    this.address = address;
  }
}
