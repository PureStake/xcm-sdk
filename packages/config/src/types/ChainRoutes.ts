import { AnyAsset, AnyChain, Asset } from '@moonbeam-network/xcm-types';
import { getKey } from '../config.utils';
import { AssetRoute } from './AssetRoute';

export interface ChainRoutesConstructorParams {
  routes: AssetRoute[];
  chain: AnyChain;
}

export class ChainRoutes {
  readonly #routes: Map<string, AssetRoute>;

  readonly chain: AnyChain;

  constructor({ routes: assets, chain }: ChainRoutesConstructorParams) {
    this.chain = chain;
    this.#routes = new Map(
      assets.map((asset) => [
        `${asset.asset.key}-${asset.destination.key}`,
        asset,
      ]),
    );
  }

  getRoutes(): AssetRoute[] {
    return Array.from(this.#routes.values());
  }

  getAssetRoutes(keyOrAsset: string | AnyAsset): AssetRoute[] {
    const key = getKey(keyOrAsset);

    return this.getRoutes().filter((cfg) => cfg.asset.key === key);
  }

  getAssetDestinations(keyOrAsset: string | AnyAsset): AnyChain[] {
    return this.getAssetRoutes(keyOrAsset).map(
      (assetConfig) => assetConfig.destination,
    );
  }

  getDestinationAssets(keyOrChain: string | AnyChain): Asset[] {
    const key = getKey(keyOrChain);

    return this.getRoutes()
      .filter((cfg) => cfg.destination.key === key)
      .map((cfg) => cfg.asset);
  }

  getAssetRoute(
    asset: string | AnyAsset,
    destination: string | AnyChain,
  ): AssetRoute {
    const assetKey = getKey(asset);
    const destKey = getKey(destination);
    const config = this.#routes.get(`${assetKey}-${destKey}`);

    if (!config) {
      throw new Error(
        `AssetConfig for ${assetKey} and destination ${destKey} not found`,
      );
    }

    return config;
  }
}
