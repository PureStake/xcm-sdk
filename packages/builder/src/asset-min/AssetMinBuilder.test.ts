import { TypeRegistry, U128 } from '@polkadot/types';
import { AssetMinBuilder } from './AssetMinBuilder';

function balanceOf(number: number | string): U128 {
  return new U128(new TypeRegistry(), number);
}

describe('assetMinBuilder', () => {
  const asset = '<ASSET>';

  describe('assetRegistry', () => {
    describe('assetMetadatas', () => {
      const config = AssetMinBuilder()
        .assetRegistry()
        .assetMetadatas()
        .build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', () => {
        expect(config.transform({ minimalBalance: balanceOf(999) })).toBe(999n);
      });
    });

    describe('currencyMetadatas', () => {
      const config = AssetMinBuilder()
        .assetRegistry()
        .currencyMetadatas()
        .build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', () => {
        expect(config.transform({ minimalBalance: balanceOf(999) })).toBe(999n);
      });
    });
  });

  describe('assets', () => {
    describe('asset', () => {
      const config = AssetMinBuilder().assets().asset().build({ asset });

      it('should be correct config', () => {
        expect(config).toMatchSnapshot();
      });

      it('should transform correctly', () => {
        expect(config.transform({ minBalance: balanceOf(999) })).toBe(999n);
      });
    });
  });
});
