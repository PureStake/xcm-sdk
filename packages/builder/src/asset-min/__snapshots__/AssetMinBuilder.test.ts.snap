// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assetMinBuilder > assetRegistry > assetMetadatas > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "assetMetadatas",
  "module": "assetRegistry",
  "transform": [Function],
  "type": "Substrate",
}
`;

exports[`assetMinBuilder > assetRegistry > assetMetadatas > should transform correctly 1`] = `999n`;

exports[`assetMinBuilder > assetRegistry > currencyMetadatas > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "currencyMetadatas",
  "module": "assetRegistry",
  "transform": [Function],
  "type": "Substrate",
}
`;

exports[`assetMinBuilder > assetRegistry > currencyMetadatas > should transform correctly 1`] = `999n`;

exports[`assetMinBuilder > assets > asset > should be correct config 1`] = `
SubstrateQueryConfig {
  "args": [
    "<ASSET>",
  ],
  "func": "asset",
  "module": "assets",
  "transform": [Function],
  "type": "Substrate",
}
`;

exports[`assetMinBuilder > assets > asset > should transform correctly 1`] = `999n`;
